File: js/euler32.js
function euler32(){
    var products = new Set();
    // pandigital product identity has 9 digits in total (1-9)
    // 2-digit * 3-digit = 4-digit (9 digits total)
    // 11 and 99 have repeating digits
    // 100-123 and 988-999 have repeating digits or 0
    for (let a = 12; a <= 98; a++){
        for (let b = 123; b <= 987; b++){
            let iden = "" + a + b + (a * b);
            // split string into character array, sort digits, rejoin into string
            let digits = iden.split("").sort().join("");
            if (digits === "123456789"){
                products.add(a * b);
            }
        }
    }
    // 1-digit * 4-digit = 4-digit (9 digits total)
    // 1000-1233 and 9877-9999 have repeating digits or 0
    for (let a = 2; a <= 9; a++){
        for (let b = 1234; b <= 9876; b++){
            let iden = "" + a + b + (a * b);
            let digits = iden.split("").sort().join("");
            if (digits === "123456789"){
                products.add(a * b);
            }
        }
    }
    var res = 0;
    for (let product of products) res += product;
    return res;
}

console.log(euler32());

File: go/euler32.go
package main
import (
    "fmt"
    "strings"
    "strconv"
)

func isPandigital(s string) bool {
    if len(s) != 9 || strings.Contains(s, "0") {
        return false
    }
    digits := make(map[rune]bool)
    for _, ch := range s {
        if digits[ch] {
            return false
        }
        digits[ch] = true
    }
    
    return len(digits) == 9
}

func main() {
    products := make(map[int]bool)
    // 2-digit * 3-digit = 4-digit
    for i := 12; i <= 98; i++ {
        for j := 123; j <= 987; j++ {
            if i * j < 10000 {
                product := i * j
                digits := strconv.Itoa(i) + strconv.Itoa(j) + strconv.Itoa(product)
                if isPandigital(digits){
                    products[product] = true
                }
            }
        }
    }
    
    // 1-digit * 4-digit = 4-digit
    for i := 1; i < 10; i++ {
        for j := 1234; j <= 9876; j++ {
            if i * j < 10000 {
                product := i * j
                digits := strconv.Itoa(i) + strconv.Itoa(j) + strconv.Itoa(product)
                if isPandigital(digits){
                    products[product] = true
                }
            }
        }
    }
    
    res := 0
    for prod := range products {
        res += prod
    }
    fmt.Println(res)
}
